File: src/main/java/com/budgetbuddy/service/CategoryOverspendingService.java
package com.budgetbuddy.service;

import com.budgetbuddy.model.CategoryOverspendingAlert;
import com.budgetbuddy.model.Transaction;
import com.budgetbuddy.model.User;
import com.budgetbuddy.repository.CategoryOverspendingAlertRepository;
import com.budgetbuddy.repository.TransactionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class CategoryOverspendingService {

    private static final Logger logger = LoggerFactory.getLogger(CategoryOverspendingService.class);

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private CategoryOverspendingAlertRepository alertRepository;

    /**
     * Detect category-level overspending for current month
     */
    public List<CategoryOverspendingAlert> detectOverspending(User user) {
        logger.info("Detecting category overspending for user: {}", user.getId());

        YearMonth currentMonth = YearMonth.now();
        LocalDate monthStart = currentMonth.atDay(1);
        LocalDate monthEnd = currentMonth.atEndOfMonth();
        LocalDate today = LocalDate.now();
        int daysElapsed = today.getDayOfMonth();
        int totalDaysInMonth = currentMonth.lengthOfMonth();

        // Get current month transactions
        List<Transaction> currentMonthTransactions = transactionRepository.findAll()
            .stream()
            .filter(t -> t.getUser().getId().equals(user.getId()))
            .filter(t -> t.getAmount() != null && t.getAmount() < 0) // Only expenses
            .filter(t -> !t.getDate().isBefore(monthStart) && !t.getDate().isAfter(monthEnd))
            .collect(Collectors.toList());

        // Get historical transactions (last 3-6 months)
        LocalDate historicalStart = LocalDate.now().minusMonths(6);
        List<Transaction> historicalTransactions = transactionRepository.findAll()
            .stream()
            .filter(t -> t.getUser().getId().equals(user.getId()))
            .filter(t -> t.getAmount() != null && t.getAmount() < 0)
            .filter(t -> t.getDate().isAfter(historicalStart) && t.getDate().isBefore(monthStart))
            .collect(Collectors.toList());

        // Group by category
        Map<String, List<Transaction>> currentByCategory = currentMonthTransactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(Transaction::getPredictedCategory));

        Map<String, List<Transaction>> historicalByCategory = historicalTransactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(Transaction::getPredictedCategory));

        List<CategoryOverspendingAlert> alerts = new ArrayList<>();

        for (Map.Entry<String, List<Transaction>> entry : currentByCategory.entrySet()) {
            String category = entry.getKey();
            List<Transaction> currentTxs = entry.getValue();

            // Calculate current month spending
            double currentAmount = currentTxs.stream()
                .mapToDouble(t -> Math.abs(t.getAmount()))
                .sum();

            // Get historical data for this category
            List<Transaction> historicalTxs = historicalByCategory.getOrDefault(category, Collections.emptyList());

            if (historicalTxs.isEmpty()) {
                continue; // Skip if no historical data
            }

            // Calculate historical average and standard deviation
            Map<YearMonth, Double> monthlySpending = historicalTxs.stream()
                .collect(Collectors.groupingBy(
                    t -> YearMonth.from(t.getDate()),
                    Collectors.summingDouble(t -> Math.abs(t.getAmount()))
                ));

            if (monthlySpending.size() < 2) {
                continue; // Need at least 2 months of data
            }

            double historicalAvg = monthlySpending.values().stream()
                .mapToDouble(Double::doubleValue)
                .average()
                .orElse(0.0);

            double[] amounts = monthlySpending.values().stream()
                .mapToDouble(Double::doubleValue)
                .toArray();

            double stdDev = calculateStandardDeviation(amounts, historicalAvg);

            // Calculate percentage increase
            double percentageIncrease = historicalAvg > 0 
                ? ((currentAmount - historicalAvg) / historicalAvg) * 100 
                : 0.0;

            // Project end-of-month total
            double projectedMonthly = (currentAmount / daysElapsed) * totalDaysInMonth;

            // Determine alert level
            CategoryOverspendingAlert.AlertLevel alertLevel = determineAlertLevel(
                percentageIncrease, currentAmount, historicalAvg, stdDev
            );

            // Only create alert if it's medium risk or higher
            if (alertLevel.ordinal() >= CategoryOverspendingAlert.AlertLevel.MEDIUM.ordinal()) {
                CategoryOverspendingAlert alert = new CategoryOverspendingAlert();
                alert.setUser(user);
                alert.setCategory(category);
                alert.setAlertLevel(alertLevel);
                alert.setCurrentAmount(currentAmount);
                alert.setHistoricalAvg(historicalAvg);
                alert.setStandardDeviation(stdDev);
                alert.setPercentageIncrease(percentageIncrease);
                alert.setProjectedMonthly(projectedMonthly);
                alert.setMonth(currentMonth);
                alert.setDaysElapsed(daysElapsed);
                alert.setIsActive(true);

                alerts.add(alert);
            }
        }

        // Save alerts
        for (CategoryOverspendingAlert alert : alerts) {
            alertRepository.save(alert);
        }

        logger.info("Detected {} category overspending alerts for user {}", alerts.size(), user.getId());
        return alerts;
    }

    /**
     * Determine alert level based on percentage increase and standard deviation
     */
    private CategoryOverspendingAlert.AlertLevel determineAlertLevel(
        double percentageIncrease, double currentAmount, double historicalAvg, double stdDev
    ) {
        // Critical: > 50% increase OR exceeds 2× standard deviation
        if (percentageIncrease > 50 || currentAmount > historicalAvg + (2 * stdDev)) {
            return CategoryOverspendingAlert.AlertLevel.CRITICAL;
        }
        // High: 25-50% increase
        if (percentageIncrease > 25) {
            return CategoryOverspendingAlert.AlertLevel.HIGH;
        }
        // Medium: 10-25% increase
        if (percentageIncrease > 10) {
            return CategoryOverspendingAlert.AlertLevel.MEDIUM;
        }
        // Low: < 10% increase
        return CategoryOverspendingAlert.AlertLevel.LOW;
    }

    private double calculateStandardDeviation(double[] values, double mean) {
        double sumSquaredDiff = 0;
        for (double value : values) {
            sumSquaredDiff += Math.pow(value - mean, 2);
        }
        return Math.sqrt(sumSquaredDiff / values.length);
    }

    /**
     * Get active alerts for a user
     */
    public List<CategoryOverspendingAlert> getActiveAlerts(User user) {
        return alertRepository.findByUserAndIsActiveTrue(user);
    }

    /**
     * Get alerts by level
     */
    public List<CategoryOverspendingAlert> getAlertsByLevel(
        User user, CategoryOverspendingAlert.AlertLevel level
    ) {
        return alertRepository.findByUserAndAlertLevel(user, level);
    }
}

File: mobile-version/app/src/main/java/com/budgetbuddy/mobile/service/CategoryOverspendingService.kt
package com.budgetbuddy.mobile.service

import com.budgetbuddy.mobile.data.dao.CategoryOverspendingAlertDao
import com.budgetbuddy.mobile.data.dao.TransactionDao
import com.budgetbuddy.mobile.data.model.CategoryOverspendingAlert
import com.budgetbuddy.mobile.data.model.Transaction
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import java.time.LocalDate
import java.time.YearMonth
import kotlin.math.abs
import kotlin.math.sqrt

class CategoryOverspendingService(
    private val transactionDao: TransactionDao,
    private val alertDao: CategoryOverspendingAlertDao
) {
    
    /**
     * Detect category-level overspending for current month
     */
    suspend fun detectOverspending(userId: Long): List<CategoryOverspendingAlert> {
        val currentMonth = YearMonth.now()
        val monthStart = currentMonth.atDay(1)
        val monthEnd = currentMonth.atEndOfMonth()
        val today = LocalDate.now()
        val daysElapsed = today.dayOfMonth
        val totalDaysInMonth = currentMonth.lengthOfMonth()
        
        // Get current month transactions
        val allTransactions = transactionDao.getTransactionsByUser(userId).first()
        val currentMonthTransactions = allTransactions
            .filter { it.userId == userId }
            .filter { it.amount < 0 }
            .filter { it.date.isAfter(monthStart.minusDays(1)) && it.date.isBefore(monthEnd.plusDays(1)) }
        
        // Get historical transactions (last 6 months)
        val historicalStart = LocalDate.now().minusMonths(6)
        val historicalTransactions = allTransactions
            .filter { it.userId == userId }
            .filter { it.amount < 0 }
            .filter { it.date.isAfter(historicalStart) && it.date.isBefore(monthStart) }
        
        // Group by category
        val currentByCategory = currentMonthTransactions
            .filter { it.predictedCategory != null }
            .groupBy { it.predictedCategory!! }
        
        val historicalByCategory = historicalTransactions
            .filter { it.predictedCategory != null }
            .groupBy { it.predictedCategory!! }
        
        val alerts = mutableListOf<CategoryOverspendingAlert>()
        
        for ((category, currentTxs) in currentByCategory) {
            // Calculate current month spending
            val currentAmount = currentTxs.sumOf { abs(it.amount) }
            
            // Get historical data for this category
            val historicalTxs = historicalByCategory[category] ?: emptyList()
            
            if (historicalTxs.isEmpty()) {
                continue // Skip if no historical data
            }
            
            // Calculate historical average and standard deviation
            val monthlySpending = historicalTxs
                .groupBy { YearMonth.from(it.date) }
                .mapValues { (_, txs) -> txs.sumOf { abs(it.amount) } }
            
            if (monthlySpending.size < 2) {
                continue // Need at least 2 months of data
            }
            
            val historicalAvg = monthlySpending.values.average()
            val variance = monthlySpending.values.map { 
                val diff = it - historicalAvg
                diff * diff
            }.average()
            val standardDeviation = sqrt(variance)
            
            // Project monthly spending based on current rate
            val projectedMonthly = (currentAmount / daysElapsed) * totalDaysInMonth
            
            // Calculate percentage increase
            val percentageIncrease = if (historicalAvg > 0) {
                ((projectedMonthly - historicalAvg) / historicalAvg) * 100
            } else {
                0.0
            }
            
            // Determine alert level
            val alertLevel = when {
                percentageIncrease > 50 || projectedMonthly > historicalAvg + (2 * standardDeviation) -> 
                    CategoryOverspendingAlert.AlertLevel.CRITICAL
                percentageIncrease > 25 -> CategoryOverspendingAlert.AlertLevel.HIGH
                percentageIncrease > 10 -> CategoryOverspendingAlert.AlertLevel.MEDIUM
                else -> CategoryOverspendingAlert.AlertLevel.LOW
            }
            
            // Check if alert already exists for this category and month
            val existingAlert = alertDao.findByCategoryAndMonth(userId, category, currentMonth)
            
            if (existingAlert == null) {
                // Create new alert
                val alert = CategoryOverspendingAlert(
                    userId = userId,
                    category = category,
                    alertLevel = alertLevel,
                    currentAmount = currentAmount,
                    historicalAvg = historicalAvg,
                    standardDeviation = standardDeviation,
                    percentageIncrease = percentageIncrease,
                    projectedMonthly = projectedMonthly,
                    month = currentMonth,
                    daysElapsed = daysElapsed,
                    isActive = true
                )
                alerts.add(alert)
            } else {
                // Update existing alert
                val updatedAlert = existingAlert.copy(
                    alertLevel = alertLevel,
                    currentAmount = currentAmount,
                    historicalAvg = historicalAvg,
                    standardDeviation = standardDeviation,
                    percentageIncrease = percentageIncrease,
                    projectedMonthly = projectedMonthly,
                    daysElapsed = daysElapsed
                )
                alerts.add(updatedAlert)
            }
        }
        
        // Save alerts
        if (alerts.isNotEmpty()) {
            alertDao.insertAll(alerts)
        }
        
        return alerts
    }
    
    /**
     * Get active alerts for user
     */
    fun getActiveAlerts(userId: Long): Flow<List<CategoryOverspendingAlert>> {
        return alertDao.getActiveAlerts(userId)
    }
    
    suspend fun getActiveAlertsSync(userId: Long): List<CategoryOverspendingAlert> {
        return alertDao.getActiveAlertsSync(userId)
    }
}

